### PR TITLE
fix: delayed engine initialization in sign unless active sessions to …

### PR DESCRIPTION
**Pull request by JonathanConn**
PoC for optimizing sign client (NOT READY FOR MERGE)

# Description

Resolves # ([browse](https://github.com/WalletConnect/walletconnect-monorepo/issues) or [create](https://github.com/WalletConnect/walletconnect-monorepo/issues/new/choose))

## Due Diligence

- [ ] Breaking change
- [ ] Requires a documentation update
  - [ ] Related docs issue/PR (if docs are not included in this PR):
- [ ] Requires a e2e/integration test update

### CODE REVIEW

File: packages/sign-client/src/client.ts [hunk 1]
```diff
@@ -85,6 +87,9 @@ export class SignClient extends ISignClient {
 
   public connect: ISignClient["connect"] = async (params) => {
     try {
+      if (!this.engineInitialized) {
```
Comment (arein) [inline packages/sign-client/src/client.ts]: nit: maybe a comment here for posterity

Or over the variable name

File: packages/sign-client/src/client.ts [hunk 2]
```diff
@@ -209,7 +214,12 @@ export class SignClient extends ISignClient {
       await this.session.init();
       await this.proposal.init();
       await this.pendingRequest.init();
-      await this.engine.init();
+
+      if (this.session.getAll().length > 0) {
```
Comment (arein) [inline packages/sign-client/src/client.ts]: What does this mean?
nit: move into variable with a name to depict why you're checking this

**Review by arein — COMMENTED** [review]
Nice! 

I know this is a draft
But how can this be tested?
Both manually and maybe in integ Tests?
Might be too hard
Just curious

File: packages/sign-client/src/client.ts (hunk 2 above)
Comment (JonathanConn) [reply packages/sign-client/src/client.ts]: I'll move it to a var so its more clear. When `session` is initialized, it loads the current sessions in, using `restore()`. I am checking to see if there are any active sessions

**Comment by JonathanConn** [comment]
Need to consider 
- `find()` and `getPendingSessionRequests()` in client because they are not async
- testing for this, ideally with performance to create a before/after comparison

**Review by xzilja — COMMENTED** [review]
LGTM 👍 will defer to Gancho for final approve

File: packages/sign-client/src/client.ts [hunk 3]
```diff
@@ -218,4 +239,17 @@ export class SignClient extends ISignClient {
       throw error;
     }
   }
+
+  // if engine was not initialized during client initialization, initialize now
+  private async initEngine() {
+    if (!this.engineInitialized) {
```
Comment (arein) [inline packages/sign-client/src/client.ts]: nitpick: this allows for race conditions. Not sure if it matters?

**Review by ganchoradkov — CHANGES_REQUESTED** [review]
I would use a different approach instead of delaying the engine init, let me explain in private

**Pull request closed without merging**